packages/google_sign_in/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java

```diff
@@ -148,372 +103,379 @@ public void onMethodCall(MethodCall call, Result result) {
-      pendingOperation = new PendingOperation(currentMethod, result);
-      return false;
+  public static final class Delegate {
+    private static final int REQUEST_CODE = 53293;
+    private static final int REQUEST_CODE_RESOLVE_ERROR = 1001;
+
+    private static final String ERROR_REASON_EXCEPTION = "exception";
+    private static final String ERROR_REASON_STATUS = "status";
+    private static final String ERROR_REASON_OPERATION_IN_PROGRESS = "operation_in_progress";
+    private static final String ERROR_REASON_CONNECTION_FAILED = "connection_failed";
+
+    private static final String STATE_RESOLVING_ERROR = "resolving_error";
+
+    private final Activity activity;
+    private final Handler handler = new Handler();
+    private final BackgroundTaskRunner backgroundTaskRunner = new BackgroundTaskRunner(1);
+
+    private boolean resolvingError = false; // Whether we are currently resolving a sign-in error
+    private GoogleApiClient googleApiClient;
+    private List<String> requestedScopes;
+    private PendingOperation pendingOperation;
+
+    public Delegate(PluginRegistry.Registrar registrar) {
+      activity = registrar.activity();
+      activity.getApplication().registerActivityLifecycleCallbacks(handler);
+      registrar.addActivityResultListener(handler);
     }
 
-    if (pendingOperation.method.equals(currentMethod)) {
-      // This method is already being handled
-      pendingOperation.resultQueue.add(result);
-    } else {
-      // Only one type of operation can be in progress at a time
-      result.error(ERROR_REASON_OPERATION_IN_PROGRESS, pendingOperation.method, null);
+    private void checkAndSetPendingOperation(String method, Result result) {
+      if (pendingOperation != null) {
+        throw new IllegalStateException(
+            "Concurrent operations detected: " + pendingOperation.method + ", " + method);
+      }
+      pendingOperation = new PendingOperation(method, result);
     }
 
-    return true;
-  }
+    /**
+     * Initializes this delegate so that it is ready to perform other operations. The Dart code
+     * guarantees that this will be called and completed before any other methods are invoked.
+     */
+    public void init(Result result, List<String> requestedScopes, String hostedDomain) {
+      // We're not initialized until we receive `onConnected`.
+      // If initialization fails, we'll receive `onConnectionFailed`
+      checkAndSetPendingOperation(METHOD_INIT, result);
 
-  /**
-   * Returns the account information for the user who is signed in to this app. If no user is signed
-   * in, tries to sign the user in without displaying any user interface.
-   */
-  private void signInSilently(Result result) {
-    if (checkAndSetPendingOperation(METHOD_SIGN_IN, result)) {
-      return;
-    }
+      try {
+        if (googleApiClient != null) {
+          // This can happen if the scopes change, or a full restart or hot reload
+          googleApiClient = null;
```

Comment:
Vestigial bit of code, no doubt (this PR just moved around existing code).  I'll remove it.